Define methods using anonymous modules and `method_missing`

kellyredding opened the pull request:
## The Goal

To alter the design to no longer explicitly define raw methods onto objects using NsOptions.

Before root-level namespace methods and root-level proxy methods would be defined directly onto the receiver of NsOptions calls.  In addition, namespaces would define raw methods on themselves as options and sub-namespaces were defined.
## The Solution

Now, root-level namespace and proxy methods are defined on anonymous modules and those anonymous modules are mixed into the receiver.  Namespace instances themselves use `method_missing` magic to expose their options and sub-namespaces.
## The Result

There is no change to the end-user experience.  However, user's will be able to extend NsOptions receivers by overwriting root-level namespaces and proxy methods and calling their `super`.  In addition, users will be able to mix in extensions to these methods.

kellyredding committed:
reworked all namespace handling classes and behavior

Each namespace has a data class to manage the namespaces child
options and namespaces.  This keeps those mgmt methods off of the
ns class and allows for more available option names.

I removed the dynamic defining on methods as well.  Now namespace
objects use method-missing magic for option/namespace readers and
writers.  Root level namespace methods are mixed in via an
anonymous module.  This allows you to overwrite readers and writers
and still be able to super up to them.

I removed all notions of a namespace tracking its 'key' and 'parent'.
Namespaces are not built simply using a name value.

kellyredding committed:
removing dynamic root method definition stuff

No longer defining actual methods on the namepace objects themselves,
and so naturally we don't want to actually define root namespace
methods either.  This moves all the root method handling into its own
class.  The class creates anonymous modules and extends or mixes them
in appropriately.  This allows subclasses to overwrite the behavior
or method-chain the root methods if needed.

In addition, these changes negate the need for the helper methods that
do the method definition stuff.  They have been removed.

Finally, the `has_options` mixin has been removed and its logic moved
to the main `ns_options` module.

Closes #48.

kellyredding committed:
breaking up the advisor into its separate concerns

This adds validation to the RootMethods handler and adds
child namespace and option checking to the namespace class.  The
checking is done by the NamespaceAdvisor class.

These are the two concerns handled by the original advisor so it is
no longer needed.

kellyredding committed:
updated the proxy handler to work in the new design

Similar to the RootMethods handler, I had to add a ProxyMethod
handler to define the proxy option/namespace methods at the root
of the proxy object.

kellyredding committed:
Updates to the remaining unit tests

I had to tweak some things to get the Assert Macro unit test passing
and I also renamed the remaining unit test files to conform with the
others.

kellyredding committed:
namespace should define on itself if it is called with a block

This behavior was accidentally dropped with the previous reworks and
was uncovered when running through the system tests.  Thanks to
@jcredding for putting in good system tests.

kellyredding committed:
tweaks to make sure all the system tests are passing

This is mostly just reformats and cleanups of the sys tests.  None
of the tests were actually changed (outside of some legacy deprecations)
 - which validates they were good system tests (thanks @jcredding).

kellyredding committed:
making sure a namespace responds to dynamic option value writers

This was accidentally dropped in the previous reworks.  It is required
by the proxy for the dynamic writer logic to actually work.

kellyredding commented:
@jcredding - yo!

kellyredding commented:
@jcredding - btw, the system tests were key here.  Thanks for putting the original work into them - helped me find a few sneaky bugs in the end.

I didn't alter them much except to remove legacy behavior.

jcredding commented:
@kellyredding - Looked through all the code (besides all the tests), everything looks real good. Nicely done sir.

kellyredding committed:
updated the README and supporting files

Tried to reduce the verbosity of the README and make it more digestable.
This also updates all the supporting files.

Closes #32.

kellyredding commented:
@jcredding - thanks!

I just updated the readme and tried to clean it up and make it less verbose (more digestable).  Will you please read over and make sure it still makes sense and conveys enough to be meaningful?  Thanks man.

That README update should be good to cover #32 for now.

jcredding commented on README.md:
```diff
@@ -108,343 +67,227 @@ module App
+
+App.settings.root    #=> "/path/to/project"
+App.settings.stage   #=> "development"
+App.settings.new_val #=> 1
```
@kellyredding - this doesn't match with the `apply` above

jcredding commented on README.md:
```diff
@@ -108,343 +67,227 @@ module App
-  option :else, :default => Proc.new { self.something }
-end
+# here, type coercion is performed
+# the is the equivalent of doing: `settings.opt2 = MyCustomTypeClass.new(789)`
```
@kellyredding - should be 'this is the equivalent...'

jcredding commented on README.md:
```diff
@@ -108,343 +67,227 @@ module App
-
-`Symbol`, `Hash` and `Array` work, but ruby doesn't provide built in type casting for these.
+settings do
+  option :opt1, MyCustomTypeClass, :args => lambda{ ['arg 1's value', 'arg 2's value'] }
```
@kellyredding - The strings for arg 1 and arg 2 value need to either mix `"` and `'` or escape the inner apostrophes.

jcredding commented:
@kellyredding - couple minor things on the README. Other then those, looks good and reads well. Doesn't feel like parts are bolted on as they did previously.

kellyredding commented:
@jcredding - thanks for that proofreading!  merging in all the things now.
